### ISSUE TITLE
Providing .pdfpc file functionality via XMP metadata 

### ISSUE THREAD

**Issue by fnevgeny** [pdfpc/pdfpc#300]
My goal is to avoid using .pdfpc files at all - at least when producing slides myself. To this end, I decided to use XMP metadata (an XML blob inside PDF). Technically, this is done by using the `hyperxmp` package (at the LaTeX side) and then parsing the relevant subset of XMP in pdfpc. An initial implementation is in the [pdfpc-latex branch in fnevgeny/pdfpc](https://github.com/fnevgeny/pdfpc/tree/pdfpc-latex). Since extending `hyperxmp` turned out to be a tricky business, I made a small `pdfpc.sty` package. One loads it like

`\usepackage[duration=15, enduserslide=10, start=14:20]{pdfpc}`

It should be self-explanatory for a pdfpc user. The `end` option is also available. In addition, I put a couple of trivial wrappers for defining notes, `\pdfpcnote` and `\pdfpcmnote`.

Ultimately, I would like to be able to use the standard `beamer`'s `\note` command, but have not figured out how to do it without hacking the `beamer` sources.

Please let me know what you think about it.

**Comment by phillipberndt** [comment]
Personally, I don't think that this is a particularly great idea.

Having two files is a minor nuisance, but it adds the benefits that come with basic text files: It's easy to add them to a VCS, and it's straightforward to manually edit them outside of pdfpc. If you embed this metadata into the PDF, on the other hand, editing notes or other metadata suddenly means recompiling the entire document.

Reducing the number of files to one is the only benefit of this change. Now that would still be a nice-to-have if there was a standard for embedding slide notes into PDFs, or at least a mechanism used by one of the big presenter programs like Acrobat or Keynote (or even Ocular/Evince). But we're talking about a proprietary mechanism that would only work with pdfpc, right?

Even then, I could live with this if it was about adding a couple of lines of code, but this is a change that adds lots of code that must eventually be maintained by someone. Even if it's simple code, this still adds complexity to the overall program.

(Regarding making this a couple of lines of code: Couldn't you store a file following exactly the pdfpc file syntax as a text node somewhere in the metadata? We could use the existing parser to handle it before processing the pdfpc file; this way, overriding the data would still be possible & generate a pdfpc file, but embedding everything into the pdf would be possible as well. And it should boil down to 2-3 commands of vala code.)

**Comment by fnevgeny** [comment]
I think I did not explain myself well. First, I did not propose to get rid of the .pdfpc files entirely. So those who are happy with them will continue to use them.

> Having two files is a minor nuisance

For me, it is a _major_ nuisance. Removing/adding a single slide requires hand editing the entire .pdfpc file: the end slide changes, all note locations shift...

> It's easy to add them to a VCS, and it's straightforward to manually edit them outside of pdfpc.

It is even easier with my approach - you have only one .tex file which is anyway under a VCS. So everything is self-contained and guaranteed to always be in sync.

> If you embed this metadata into the PDF, on the other hand, editing notes or other metadata suddenly means recompiling the entire document.

First, this is what _I_ call a minor nuisance :) Run `make` and in a few seconds it is ready. Second, no, at least for editing notes, just use your favourite PDF viewer for that. There are also [software for editing XMP](https://en.wikipedia.org/wiki/Comparison_of_metadata_editors).

> Now that would still be a nice-to-have if there was a standard for embedding slide notes into PDFs, or at least a mechanism used by one of the big presenter programs like Acrobat or Keynote (or even Ocular/Evince). But we're talking about a proprietary mechanism that would only work with pdfpc, right?

Absolutely not. The notes actually is the easiest part - these are the standard PDF annotations, the functionality is already in pdfpc And yes, you can add/edit the PDF annotations using evince/okular etc as well. XMP is used only for storing pdfpc-specific metadata - talk duration start/stop time etc. I wish there were a standard for specifying these, but to the best of my knowledge there is not.

> but this is a change that adds lots of code

Fifty lines or so, to be precise. Please have a look at the branch.

> Couldn't you store a file following exactly the pdfpc file syntax as a text node somewhere in the metadata?

Nope. In principle, it is possible, but Poppler does not have a GLib binding for getting non-standard non-XMP metadata. One can file a request, but in the best case it will take a year to propagate to distributions. But I am sure they will disagree. XMP is [_the_ standard](https://www.iso.org/news/2012/03/Ref1525.html) for adding PDF metadata.

**Comment by phillipberndt** [comment]
> First, I did not propose to get rid of the .pdfpc files entirely. So those who are happy with them will continue to use them.

Yes, I got that. Having two separate code paths for parsing different sets of configuration files with different syntax and different names for the options is what worries me, actually.

> For me, it is a major nuisance. Removing/adding a single slide requires hand editing the entire .pdfpc file: the end slide changes, all note locations shift...

That sounds as if it isn't really the two files which are the problem, but the fact that the pdfpc file isn't automatically updated/regenerated. Are you aware of the [pdfpc-latex-notes](https://github.com/cebe/pdfpc-latex-notes) project? It could be easily extended to create a full-fledged pdfpc file, including setting the options, from Latex. No changes to pdfpc would be required, and you'd still no longer have to edit the pdfpc file on your own.

> Nope. In principle, it is possible, but Poppler does not have a GLib binding for getting non-standard non-XMP metadata.

What I'm suggesting is to embed the pdfpc file into the XMP stream, properly escaped to form valid XML. E.g., as a `pdfpc:rcfile` tag containing the file as CDATA.

**Comment by fnevgeny** [comment]
> Having two separate code paths for parsing different sets of configuration files with different syntax and different names for the options is what worries me, actually.

There are already two code paths for configuration - command line and .pdfpc. With slightly different key names, BTW. I can make the XMP keywords exactly the same as in pdfpc, but it will add headache to users - escaping underscores in LaTeX. What is the point of making them absolutely identical? Or do you mean the names of the XML tags? Ideally, end-users should never have to see them.

> That sounds as if it isn't really the two files which are the problem, but the fact that the pdfpc file isn't automatically updated/regenerated.

This is number one problem, right.

> Are you aware of the pdfpc-latex-notes project?

Of course. But I am getting there without the need for an extra file. To me, it is more elegant.

> What I'm suggesting is to embed the pdfpc file into the XMP stream, properly escaped to form valid XML. E.g., as a pdfpc:rcfile tag containing the file as CDATA.

Oh, I see. But why to go an imperfect route? XML is not less  powerful than the ini-like .pdfpc syntax. The real code for parsing the options is ten lines inside `pdfpc`; the XML parser would remain there to fetch the metadata anyway. Again, XMP is the standard. Meaning there are generic utilities to deal with this kind of metadata - easier to integrate with other software, e.g., writing a plugin for `OO` etc.

**Comment by phillipberndt** [comment]
Ok, let's not discuss implementation details for now. I'm still not convinced that we need this at all.

> To me, it is more elegant. [..] But why to go an imperfect route?

That's the only indisputable reason for having this feature that you have brought up so far. Here's why I believe we should:

Because it isn't clear that you'll be around in a few years when something breaks, and any new code adds complexity and raises the entry barrier for new contributors/maintainers. Long lived projects should aim for code that is easy to maintain - that includes valuing simplicity over elegance, and weighing up the complexity added by new features against their use. I simply don't see that equation balance towards the right side here.

(In case this isn't clear, please note: I'm only giving an opinion here. Andreas might disagree, and as the principal maintainer he definitely has the final word in this regard. Also, I'm talking about the idea to integrate the pdfpc file into the pdf here, not about improving the LaTeX-integration.)

**Comment by fnevgeny** [comment]
> That's the only indisputable reason for having this feature that you have brought up so far.

OK :). But elegance is also a factor in code maintainability.

I certainly understand your points.

**Comment by AndreasBilke** [comment]
First, I think it is a valid wish to bundle the PDF and the pdfpc metadata together. I personally would not use it but that shouldn't count here.

What I don't like about the current implementation of @fnevgeny is, that we have additional parsing code for the XMP data. If we add new features to `.pdfpc` files we would have to implement them also for the XMP stuff to keep both ways in sync (and I think they should always be in sync).

@fnevgeny I have no clue what could be done on the XMP side (and in latex) but I see the point of @phillipberndt of just putting plain pdfpc text in the XMP data. One could fetch this text and throw it into the `.pdfpc` parser without any changes. It would still be possible to edit the `.pdfpc` data within the PDF with any XMP editor (since they just XML editors right?).

If I see it correctly, non of the pdf viewers (or presentation programs) support this kind of configuration so we would be in any case (real XMP or plain text within XMP) a "proprietary"  solution.

I generally favour the pragmatic solution over the elegant one. We had for a long time very elegant code in pdfpc which led us in a situation where the code was mostly unmaintainable.

**Comment by fnevgeny** [comment]
@AndreasBilke Just to make sure we are talking about the same thing: Does this chunk of code

```
foreach (var entry in tags.entries) {
    switch (entry.key) {
        case "Duration":
            set_duration(entry.value.to_int());
            break;
        case "EndUserSlide":
            set_end_user_slide(entry.value.to_int());
            break;
        case "StartTime":
            Options.start_time = entry.value;
            break;
        case "EndTime":
            Options.end_time = entry.value;
            break;
        default:
            GLib.printerr("unknown XMP entry %s\n", entry.key);
            break;
}
```

really worry you as becoming unmaintainable in the future? Even with several more hypothetical extra `case`'s?

Most of the current .pdfpc contents is simply not needed for embedding in PDF: notes (they are real PDF annotations) and "skips" (beamer overlays work nicely with pdfpc without this ad-hoc workaround). So embedding/parsing a tiny subset of .pdfpc (just the few key/value pairs above) as a text blob would anyway need code branching.

**Comment by AndreasBilke** [comment]
@fnevgeny Yes, I meant that code. And yes I worry about future changes. Would it be possible (technically) to dump plain text to XMP data? How complicated would it be to do this on the latex side?

**Comment by fnevgeny** [comment]
> Would it be possible (technically) to dump plain text to XMP data?

Yes, it should be possible.

> How complicated would it be to do this on the latex side?

I believe not difficult. Just ugly. And (if the functionality will be extended) this ugliness will only grow. Say,  on my personal pdfpc TODO list is to add per-slide time duration (not necessarily for all slides, only for too complex ones) to use it for calculating progress with the timer. Or introduce "profiles" (that is, different settings for, e.g., rehearsing/real presentation). XML is well suited for rich structured data. The .pdfpc format is not.

As a compromise, I suggest adding a "pdfpc" flag to the pdfpc.sty package options, in which case everything would go to a .pdfpc file instead of embedding in PDF. This way, if the burden of maintaining that `switch` becomes backbreaking, you just remove it and tell LaTeX users to use this flag (or just enable it by default).

@phillipberndt, @AndreasBilke Deal?

**Comment by AndreasBilke** [comment]
@fnevgeny I am not fully convinced but I would give this a chance. So: Deal :-)

**Comment by AdrianVollmer** [comment]
I also like this idea. Just a minor comment:

> Ultimately, I would like to be able to use the standard beamer's \note command, but have not figured out how to do it without hacking the beamer sources.

Can't you just use `\renewcommand`? In general, it's not too hard to redefine commands in LaTeX, unless I'm overlooking some particularity here.

**Comment by fnevgeny** [comment]
@AdrianVollmer Frankly, I don't remember what I had in mind ~2 years ago...

**Comment by AndreasBilke** [comment]
@fnevgeny do you want to keep this issue open then?

**Comment by fnevgeny** [comment]
> @fnevgeny do you want to keep this issue open then?

Yes. My comment was only re: modifying the \note command in beamer.

**Comment by fnevgeny** [comment]
> Can't you just use `\renewcommand`? In general, it's not too hard to redefine commands in LaTeX, unless I'm overlooking some particularity here.

@AdrianVollmer OK, now I recall. `\note` is much more than a simple comment. For one, it may have `<⟨overlay specification⟩>` (this is not currently supported by pdfpc, but there is no principal problem; #239). Second, it is affected by quite some other beamer commands and options, e.g., `\AtBeginNote`.

**Comment by AndreasBilke** [comment]
Fixed with #471
